### DISCUSSION TITLE
Unable to run xplat on WSL

### DISCUSSION THREAD

**Discussion by curia-damiano** [AvaloniaUI/Avalonia#11487] [Q&A] [answered]
**Steps to Reproduce**
I have created a new app following https://docs.avaloniaui.net/tutorials/developing-for-mobile/create-a-cross-platform-solution:
- dotnet new install avalonia.templates
- dotnet new avalonia.xplat --name AvaloniaCross
The app gets created and it compiles successfully. It also runs on Windows.

I install a clean WSL image, simply with:
- wsl --install

When I choose to run on WSL, Visual Studio complains that Ubuntu is missing dot.net.
I let it to install, but then I get this error:

<img width="800" alt="Avalonia_3" src="https://github.com/AvaloniaUI/Avalonia/assets/18611507/b58d1509-23b7-4e72-a1db-2de3cc89add9">

Reading various issues that it is related to possible issues in the .NET installation, I completely unregister Ubuntu and create a new instance.
I install .NET 7 following this guide: https://learn.microsoft.com/en-us/dotnet/core/install/linux-ubuntu#register-the-microsoft-package-repository
(plus "sudo apt install dotnet-sdk-7.0)

When I try to run the project from Visual Studio, it automatically runs a script to configure the debugger.

But then I get the same error.

I check the SDKs installed (I get no-one, strangely) but I get the runtime (that I think should be enough)

<img width="867" alt="Ubuntu" src="https://github.com/AvaloniaUI/Avalonia/assets/18611507/2d2d6366-6201-4241-b424-999eff265f3b">

Because I have also installed Visual Studio preview, I have also added a global.json to the solution, forcing the SDK to 7.0.302, and then recompiled the application, but still no success.

What am I missing to run a simply Avalonia UI in WSL?

**Desktop (please complete the following information):**
Windows 11, latest stable version as of today
WSL 2
All Windows updates installed, all Visual Studio updates installed
Tried on both VS 15.6 and 15.7 beta (with the global.json to force the use of the .NET 7 SDK)

**Comment by curia-damiano** [comment]
Update: I have also tried following https://www.appsloveworld.com/csharp/100/1647/type-initializer-for-skiasharp-skimageinfo-threw-an-exception-avalonia, but same result :(

**Comment by timunie** [comment]
Does it work without using Visual Studio? I mean compile the App and then open directly from WSL?

**Comment by curia-damiano** [comment]
Stupid question: when I run the app from Visual Studio, is it deployed on WSL?
I can't find anything (related to my app) inside /bin or /usr/bin

**Reply by maxkatz6** [reply]
Depending on the project launch profile. But I haven't used WSL from Visual Studio.
You can publish the app, mark it as executable and send it to the WSL directory and run it from there, just to test. Or "dotnet run" from the inside of WSL.
Or, you can install VSCode and configure remote debug with WSL https://code.visualstudio.com/docs/remote/wsl. This way you will be able to edit code and launch it with "dotnet run" as well.

**Comment by maxkatz6** [comment]
Does it use ubuntu snap? If so, snap has known problems with dependencies of that kind.

Relevant issues:
https://github.com/AvaloniaUI/Avalonia/issues/5190
https://github.com/AvaloniaUI/Avalonia/issues/3679

**Comment by grokys** [comment, marked as answer]
You need to make sure a few dependencies are installed:

```
sudo apt install libice6
sudo apt install libsm6
sudo apt install libfontconfig1
```

**Comment by curia-damiano** [comment]
Thank you @grokys , it worked!!!

If I search in Google for "avalonia ui" libice6 libsm6 libconfconfig1, I don't get any result.
Are these dependencies documented anywhere?
Anyway I have blogged the experience [here](https://curia.me/develop-graphical-linux-apps-with-avalonia-ui/).

Plus: searching for Avalonia and the Skia type initialization error, I got redirected only to issues with wrong installation of the .NET framework.

**Reply by maxkatz6** [reply]
Would be great to add this blogpost to this page https://github.com/AvaloniaCommunity/awesome-avalonia

**Reply by maroonedsia** [reply]
Thank you @curia-damiano for bringing this up! 👍 

@maxkatz6 I have gone up and down over the docs and couldn't make a GUI app work on WSL. These instructions sound to be essential / prerequisite for setting up the environment for developing Linux GUI apps on Windows. So, I hope the official docs can be updated with these (instead of putting it on Awesome Avaliona)... Thanks! 👍

**Reply by jp2masa** [reply]
The FAQ in the docs was updated recently ([ref](https://github.com/AvaloniaUI/Avalonia/discussions/13771#discussioncomment-7702016)) to describe the supported platforms more accurately. It includes a list of prerequisites for WSL 2: https://docs.avaloniaui.net/docs/faq#linux-distros.

**Reply by maxkatz6** [reply]
@maroonedsia yes, there is nothing special about WSL, except these dependencies which we already mentioned in FAQ.

Alternatively, @curia-damiano guide can be transferred to our documentation, if @curia-damiano is fine with that.

**Reply by curia-damiano** [reply]
Hi @maxkatz6 ,
I am totally happy to help and you can use my guide in your documentation.

**Reply by timunie** [reply]
If anyone wants to prepare a PR for our docs, find the repo here:  https://github.com/AvaloniaUI/Avalonia-docs

**Reply by maroonedsia** [reply]
Thank you all! It'll be awesome if @curia-damiano creates the PR. Cheers! 😎

**Comment by cemalgulbeyaz** [comment]
I have fixed Debug problem with this guide thank you so much. But there is another issue after start to debug: there some visual corrupts on GUI, first it think about graphic driver and i installed it on WSL then still same problem. I have recorded and upload problem video on youtube:
[https://youtu.be/dSxpSyLP6a0](url)

if it works i try to upload here too

https://github.com/AvaloniaUI/Avalonia/assets/12149285/0551d6a1-fea0-46fd-bd95-a861ba1a117a

Even I have no idea how can search this problem. Which keywords i need to search it?
